builtin/pure_ysh.py

```diff
@@ -102,6 +102,106 @@ def Run(self, cmd_val):
         return 0
 
 
+class Ctx(vm._Builtin):
+
+    def __init__(self, mem, cmd_ev):
+        # type: (state.Mem, CommandEvaluator) -> None
+        self.mem = mem
+        self.cmd_ev = cmd_ev  # To run blocks
+
+    def _Push(self, context, block):
+        # type: (Dict[str, value_t], command_t) -> int
+
+        # TODO: ctx_Ctx
+        self.mem.PushContextStack(context)
+        self.cmd_ev.EvalCommand(block)
+        self.mem.PopContextStack()
+
+        return 0
+
+    def _Set(self, updates):
+        # type: (Dict[str, value_t]) -> int
+
+        ctx = self.mem.GetContext()
+        if ctx is None:
+            # TODO: raise error
+            return 1
+
+        ctx.update(updates)
+
+        return 0
+
+    def _Emit(self, field, item, blame):
+        # type: (str, value_t, loc_t) -> int
+
+        ctx = self.mem.GetContext()
+        if ctx is None:
+            # TODO: raise error
+            return 1
+
+        if field not in ctx:
+            ctx[field] = value.List([])
+
+        UP_arr = ctx[field]
+        if UP_arr.tag() != value_e.List:
+            raise error.TypeErr(
+                UP_arr,
+                "Expected the context item '%s' to be a List" % (field), blame)
+
+        arr = cast(value.List, UP_arr)
+        arr.items.append(item)
+
+        return 0
+
+    def Run(self, cmd_val):
+        # type: (cmd_value.Argv) -> int
+        _, arg_r = flag_spec.ParseCmdVal('ctx',
+                                         cmd_val,
+                                         accept_typed_args=True)
+
+        verb = arg_r.Peek()
```

Comment:
BTW there is a Peek2() that gives a location that could be more specific?  I didn't test it, but if you do `ctx zzz` it could be more precise